File: backend-rust/src/util/permission_bitfield.rs
use std::collections::{HashMap, HashSet};

const V3_MENU_SHIFT: u64 = 48;
const V3_ACTION_SHIFT: u64 = 40;
const V3_SCOPE_SHIFT: u64 = 32;
const V3_MENU_MASK: u64 = 0xFFFF;
const V3_ACTION_MASK: u64 = 0xFF;
const V3_SCOPE_MASK: u64 = 0xFF;
const V3_RESERVED_SIGNATURE: u64 = 0x43534D33; // "CSM3"

const V3_ACTION_VIEW: u64 = 0;
const V3_ACTION_CREATE: u64 = 1;
const V3_ACTION_EDIT: u64 = 2;
const V3_ACTION_DELETE: u64 = 3;
const V3_ACTION_EXPORT: u64 = 4;

const V3_SCOPE_OWNER: u64 = 0;
const V3_SCOPE_DEPARTMENT: u64 = 1;
const V3_SCOPE_BRANCH: u64 = 2;
const V3_SCOPE_ALL: u64 = 3;

// Legacy flat bit layout (pre-v3) — mirrors Java PermissionBitfieldUtil
const ACTION_VIEW: u64 = 31;
const ACTION_CREATE: u64 = 33;
const ACTION_EDIT: u64 = 32;
const ACTION_DELETE: u64 = 34;
const ACTION_EXPORT: u64 = 35;
const DATA_SCOPE_OWNER: u64 = 41;
const DATA_SCOPE_DEPARTMENT: u64 = 42;
const DATA_SCOPE_BRANCH: u64 = 43;
const DATA_SCOPE_ALL: u64 = 44;

pub struct PermissionBitfieldUtil;

impl PermissionBitfieldUtil {
    pub const SCHEMA_V3: &'static str = "v3";

    pub fn build_bitfield(permissions: &[String], menus: &[String], dev: bool) -> u64 {
        let mut menu_mask = 0u64;
        let mut action_mask = 0u64;
        let mut scope_mask = 0u64;

        if dev {
            action_mask = set_mask_bit(action_mask, V3_ACTION_VIEW);
            action_mask = set_mask_bit(action_mask, V3_ACTION_CREATE);
            action_mask = set_mask_bit(action_mask, V3_ACTION_EDIT);
            action_mask = set_mask_bit(action_mask, V3_ACTION_DELETE);
            action_mask = set_mask_bit(action_mask, V3_ACTION_EXPORT);
            scope_mask = set_mask_bit(scope_mask, V3_SCOPE_ALL);
        }

        for raw in permissions {
            let token = normalize_token(raw);
            if token.is_empty() {
                continue;
            }
            match token.as_str() {
                "dev" | "admin" => {
                    action_mask = set_mask_bit(action_mask, V3_ACTION_VIEW);
                    action_mask = set_mask_bit(action_mask, V3_ACTION_CREATE);
                    action_mask = set_mask_bit(action_mask, V3_ACTION_EDIT);
                    action_mask = set_mask_bit(action_mask, V3_ACTION_DELETE);
                    action_mask = set_mask_bit(action_mask, V3_ACTION_EXPORT);
                    scope_mask = set_mask_bit(scope_mask, V3_SCOPE_ALL);
                }
                "user" => {
                    action_mask = set_mask_bit(action_mask, V3_ACTION_VIEW);
                    scope_mask = set_mask_bit(scope_mask, V3_SCOPE_OWNER);
                }
                "view" | "read" => action_mask = set_mask_bit(action_mask, V3_ACTION_VIEW),
                "create" | "add" | "insert" => {
                    action_mask = set_mask_bit(action_mask, V3_ACTION_CREATE)
                }
                "edit" | "update" | "write" => {
                    action_mask = set_mask_bit(action_mask, V3_ACTION_EDIT)
                }
                "delete" | "remove" => action_mask = set_mask_bit(action_mask, V3_ACTION_DELETE),
                "export" => action_mask = set_mask_bit(action_mask, V3_ACTION_EXPORT),
                "scope:owner" | "owner" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_OWNER),
                "scope:department" | "department" | "team" => {
                    scope_mask = set_mask_bit(scope_mask, V3_SCOPE_DEPARTMENT)
                }
                "scope:branch" | "branch" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_BRANCH),
                "scope:all" | "all" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_ALL),
                _ => {}
            }
        }

        for raw in menus {
            let token = normalize_token(raw);
            if token.is_empty() {
                continue;
            }
            if let Some(menu_index) = parse_menu_index(&token) {
                if menu_index <= 15 {
                    menu_mask = set_mask_bit(menu_mask, menu_index);
                }
            }
            match token.as_str() {
                "scope:owner" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_OWNER),
                "scope:department" | "scope:team" => {
                    scope_mask = set_mask_bit(scope_mask, V3_SCOPE_DEPARTMENT)
                }
                "scope:branch" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_BRANCH),
                "scope:all" => scope_mask = set_mask_bit(scope_mask, V3_SCOPE_ALL),
                _ => {}
            }
        }

        (menu_mask & V3_MENU_MASK) << V3_MENU_SHIFT
            | (action_mask & V3_ACTION_MASK) << V3_ACTION_SHIFT
            | (scope_mask & V3_SCOPE_MASK) << V3_SCOPE_SHIFT
            | (V3_RESERVED_SIGNATURE & 0xFFFF_FFFF)
    }

    /// Mirrors Java `toCompactToken` — base-36 uppercase.
    pub fn to_compact_token(bitfield: u64) -> String {
        to_base36_upper(bitfield)
    }

    pub fn parse_security_token(raw: Option<&str>) -> Option<u64> {
        let text = raw?.trim().replace('_', "");
        if text.is_empty() {
            return None;
        }

        if let Some(rest) = text.strip_prefix("b36:").or_else(|| text.strip_prefix("B36:")) {
            return u64::from_str_radix(rest, 36).ok().map(normalize_to_single_token);
        }

        if let Some(rest) = text.strip_prefix("b64:").or_else(|| text.strip_prefix("B64:")) {
            use base64::Engine;
            let bytes = base64::engine::general_purpose::URL_SAFE_NO_PAD
                .decode(rest)
                .or_else(|_| base64::engine::general_purpose::STANDARD.decode(rest))
                .ok()?;
            if bytes.len() != 8 {
                return None;
            }
            let mut value = 0u64;
            for b in bytes {
                value = (value << 8) | u64::from(b);
            }
            return Some(normalize_to_single_token(value));
        }

        if let Some(rest) = text.strip_prefix("0x").or_else(|| text.strip_prefix("0X")) {
            return u64::from_str_radix(rest, 16)
                .ok()
                .map(normalize_to_single_token);
        }

        if text.chars().all(|c| c.is_ascii_digit()) {
            return text
                .parse::<u64>()
                .ok()
                .map(normalize_to_single_token);
        }

        if text.chars().all(|c| c.is_ascii_alphanumeric()) {
            return u64::from_str_radix(&text, 36)
                .ok()
                .map(normalize_to_single_token);
        }

        text.parse::<u64>().ok().map(normalize_to_single_token)
    }

    pub fn has_bit(bitfield: u64, bit_index: i32) -> bool {
        has_bit_v3(normalize_to_single_token(bitfield), bit_index)
    }

    pub fn permissions_from_bitfield(raw: Option<&str>) -> Vec<String> {
        let Some(token) = Self::parse_security_token(raw) else {
            return vec![];
        };
        let mut out = Vec::new();
        for (bit, label) in ACTION_BIT_TO_TOKEN.iter() {
            if Self::has_bit(token, *bit) {
                out.push(label.to_string());
            }
        }
        match resolve_data_scope_v3(token).as_str() {
            "ALL" => out.push("scope:all".into()),
            "BRANCH" => out.push("scope:branch".into()),
            "DEPARTMENT" => out.push("scope:department".into()),
            "OWNER" => out.push("scope:owner".into()),
            _ => {}
        }
        out
    }

    pub fn menus_from_bitfield(raw: Option<&str>) -> Vec<String> {
        let Some(token) = Self::parse_security_token(raw) else {
            return vec![];
        };
        let mut out = Vec::new();
        for (bit, label) in MENU_BIT_TO_TOKEN.iter() {
            if Self::has_bit(token, *bit) {
                out.push(label.to_string());
            }
        }
        out
    }

    pub fn merge_unique_case_insensitive(base: &[String], extra: &[String]) -> Vec<String> {
        let mut merged = Vec::new();
        let mut seen = HashSet::new();
        for value in base.iter().chain(extra.iter()) {
            let normalized = value.trim();
            if normalized.is_empty() {
                continue;
            }
            let key = normalized.to_ascii_lowercase();
            if seen.insert(key) {
                merged.push(normalized.to_string());
            }
        }
        merged
    }

    pub fn resolve_data_scope(bitfield: u64) -> String {
        resolve_data_scope_v3(normalize_to_single_token(bitfield))
    }
}

fn resolve_data_scope_v3(token: u64) -> String {
    let scope_mask = extract_v3_scope_mask(token);
    if has_mask_bit(scope_mask, V3_SCOPE_ALL) {
        return "ALL".into();
    }
    if has_mask_bit(scope_mask, V3_SCOPE_BRANCH) {
        return "BRANCH".into();
    }
    if has_mask_bit(scope_mask, V3_SCOPE_DEPARTMENT) {
        return "DEPARTMENT".into();
    }
    if has_mask_bit(scope_mask, V3_SCOPE_OWNER) {
        return "OWNER".into();
    }
    "NONE".into()
}

fn is_security_token_v3(token: u64) -> bool {
    (token & 0xFFFF_FFFF) == (V3_RESERVED_SIGNATURE & 0xFFFF_FFFF)
}

fn normalize_to_single_token(token: u64) -> u64 {
    if is_security_token_v3(token) {
        return token;
    }

    let mut menu_mask = 0u64;
    for i in 0..=15 {
        if token & (1u64 << i) != 0 {
            menu_mask = set_mask_bit(menu_mask, i);
        }
    }

    let mut action_mask = 0u64;
    if token & (1u64 << ACTION_VIEW) != 0 {
        action_mask = set_mask_bit(action_mask, V3_ACTION_VIEW);
    }
    if token & (1u64 << ACTION_CREATE) != 0 {
        action_mask = set_mask_bit(action_mask, V3_ACTION_CREATE);
    }
    if token & (1u64 << ACTION_EDIT) != 0 {
        action_mask = set_mask_bit(action_mask, V3_ACTION_EDIT);
    }
    if token & (1u64 << ACTION_DELETE) != 0 {
        action_mask = set_mask_bit(action_mask, V3_ACTION_DELETE);
    }
    if token & (1u64 << ACTION_EXPORT) != 0 {
        action_mask = set_mask_bit(action_mask, V3_ACTION_EXPORT);
    }

    let mut scope_mask = 0u64;
    if token & (1u64 << DATA_SCOPE_OWNER) != 0 {
        scope_mask = set_mask_bit(scope_mask, V3_SCOPE_OWNER);
    }
    if token & (1u64 << DATA_SCOPE_DEPARTMENT) != 0 {
        scope_mask = set_mask_bit(scope_mask, V3_SCOPE_DEPARTMENT);
    }
    if token & (1u64 << DATA_SCOPE_BRANCH) != 0 {
        scope_mask = set_mask_bit(scope_mask, V3_SCOPE_BRANCH);
    }
    if token & (1u64 << DATA_SCOPE_ALL) != 0 {
        scope_mask = set_mask_bit(scope_mask, V3_SCOPE_ALL);
    }

    (menu_mask & V3_MENU_MASK) << V3_MENU_SHIFT
        | (action_mask & V3_ACTION_MASK) << V3_ACTION_SHIFT
        | (scope_mask & V3_SCOPE_MASK) << V3_SCOPE_SHIFT
        | (V3_RESERVED_SIGNATURE & 0xFFFF_FFFF)
}

fn has_bit_v3(token: u64, bit_index: i32) -> bool {
    if bit_index < 0 {
        return false;
    }
    if (0..=15).contains(&bit_index) {
        return has_mask_bit(extract_v3_menu_mask(token), bit_index as u64);
    }
    if (16..=30).contains(&bit_index) {
        return false;
    }
    match bit_index {
        x if x == ACTION_VIEW as i32 => has_mask_bit(extract_v3_action_mask(token), V3_ACTION_VIEW),
        x if x == ACTION_CREATE as i32 => {
            has_mask_bit(extract_v3_action_mask(token), V3_ACTION_CREATE)
        }
        x if x == ACTION_EDIT as i32 => has_mask_bit(extract_v3_action_mask(token), V3_ACTION_EDIT),
        x if x == ACTION_DELETE as i32 => {
            has_mask_bit(extract_v3_action_mask(token), V3_ACTION_DELETE)
        }
        x if x == ACTION_EXPORT as i32 => {
            has_mask_bit(extract_v3_action_mask(token), V3_ACTION_EXPORT)
        }
        x if x == DATA_SCOPE_OWNER as i32 => {
            has_mask_bit(extract_v3_scope_mask(token), V3_SCOPE_OWNER)
        }
        x if x == DATA_SCOPE_DEPARTMENT as i32 => {
            has_mask_bit(extract_v3_scope_mask(token), V3_SCOPE_DEPARTMENT)
        }
        x if x == DATA_SCOPE_BRANCH as i32 => {
            has_mask_bit(extract_v3_scope_mask(token), V3_SCOPE_BRANCH)
        }
        x if x == DATA_SCOPE_ALL as i32 => has_mask_bit(extract_v3_scope_mask(token), V3_SCOPE_ALL),
        _ => false,
    }
}

fn extract_v3_menu_mask(token: u64) -> u64 {
    (token >> V3_MENU_SHIFT) & V3_MENU_MASK
}

fn extract_v3_action_mask(token: u64) -> u64 {
    (token >> V3_ACTION_SHIFT) & V3_ACTION_MASK
}

fn extract_v3_scope_mask(token: u64) -> u64 {
    (token >> V3_SCOPE_SHIFT) & V3_SCOPE_MASK
}

fn set_mask_bit(mask: u64, bit_position: u64) -> u64 {
    if bit_position > 62 {
        return mask;
    }
    mask | (1u64 << bit_position)
}

fn has_mask_bit(mask: u64, bit_position: u64) -> bool {
    if bit_position > 62 {
        return false;
    }
    (mask & (1u64 << bit_position)) != 0
}

fn normalize_token(value: &str) -> String {
    value.trim().to_ascii_lowercase()
}

fn parse_menu_index(token: &str) -> Option<u64> {
    if let Ok(index) = token.parse::<u64>() {
        if index <= 30 {
            return Some(index);
        }
    }
    if let Some(index) = known_menu_bits().get(token) {
        return Some(*index);
    }
    if let Some(rest) = token.strip_prefix("menu:") {
        if let Ok(index) = rest.parse::<u64>() {
            if index <= 30 {
                return Some(index);
            }
        }
    }
    None
}

fn known_menu_bits() -> HashMap<&'static str, u64> {
    HashMap::from([
        ("dashboard", 0),
        ("/dashboard", 0),
        ("home", 0),
        ("/home", 0),
        ("user", 1),
        ("/system/user", 1),
        ("role", 2),
        ("/system/role", 2),
        ("menu", 3),
        ("/system/menu", 3),
        ("dept", 4),
        ("/system/dept", 4),
        ("developer", 5),
        ("/system/developer", 5),
        ("broadcast", 6),
        ("/system/broadcast", 6),
        ("report", 7),
        ("/system/report", 7),
        ("crm", 8),
        ("/crm", 8),
    ])
}

fn to_base36_upper(mut value: u64) -> String {
    if value == 0 {
        return "0".to_string();
    }
    const ALPHABET: &[u8; 36] = b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    let mut out = Vec::new();
    while value > 0 {
        let rem = (value % 36) as usize;
        out.push(ALPHABET[rem]);
        value /= 36;
    }
    out.iter().rev().map(|b| *b as char).collect()
}

const ACTION_BIT_TO_TOKEN: &[(i32, &str)] = &[
    (ACTION_VIEW as i32, "view"),
    (ACTION_CREATE as i32, "create"),
    (ACTION_EDIT as i32, "edit"),
    (ACTION_DELETE as i32, "delete"),
    (ACTION_EXPORT as i32, "export"),
];

const MENU_BIT_TO_TOKEN: &[(i32, &str)] = &[
    (0, "/home"),
    (1, "/system/user"),
    (2, "/system/role"),
    (3, "/system/menu"),
    (4, "/system/dept"),
    (5, "/system/developer"),
    (6, "/system/broadcast"),
    (7, "/system/report"),
    (8, "/crm"),
];

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn menu_paths_map_to_known_bits() {
        let bitfield = PermissionBitfieldUtil::build_bitfield(&[], &["/system/user".into(), "/crm".into()], false);
        assert!(PermissionBitfieldUtil::has_bit(bitfield, 1));
        assert!(PermissionBitfieldUtil::has_bit(bitfield, 8));
        assert!(!PermissionBitfieldUtil::has_bit(bitfield, 0));
    }

    #[test]
    fn compact_token_is_base36() {
        let bitfield = PermissionBitfieldUtil::build_bitfield(&["admin".into()], &[], false);
        let token = PermissionBitfieldUtil::to_compact_token(bitfield);
        assert!(token.chars().all(|c| c.is_ascii_alphanumeric()));
        assert_eq!(
            PermissionBitfieldUtil::parse_security_token(Some(&token)),
            Some(bitfield)
        );
    }

    #[test]
    fn menus_round_trip_from_bitfield() {
        let menus = vec!["/system/user".into(), "/crm".into()];
        let bitfield = PermissionBitfieldUtil::build_bitfield(&[], &menus, false);
        let token = PermissionBitfieldUtil::to_compact_token(bitfield);
        let decoded = PermissionBitfieldUtil::menus_from_bitfield(Some(&token));
        assert!(decoded.contains(&"/system/user".to_string()));
        assert!(decoded.contains(&"/crm".to_string()));
    }
}

File: backend-go/internal/util/permission_bitfield.go
package util

import (
	"encoding/base64"
	"math/big"
	"strings"
)

const (
	SchemaV3 = "v3"

	v3MenuShift = 48
	v3ActionShift = 40
	v3ScopeShift = 32
	v3MenuMask = 0xFFFF
	v3ActionMask = 0xFF
	v3ScopeMask = 0xFF
	v3ReservedSignature = 0x43534D33

	v3ActionView = 0
	v3ActionCreate = 1
	v3ActionEdit = 2
	v3ActionDelete = 3
	v3ActionExport = 4

	v3ScopeOwner = 0
	v3ScopeDepartment = 1
	v3ScopeBranch = 2
	v3ScopeAll = 3

	actionView = 31
	actionCreate = 33
	actionEdit = 32
	actionDelete = 34
	actionExport = 35
	dataScopeOwner = 41
	dataScopeDepartment = 42
	dataScopeBranch = 43
	dataScopeAll = 44
)

var menuBitToToken = map[int]string{
	0: "/home",
	1: "/system/user",
	2: "/system/role",
	3: "/system/menu",
	4: "/system/dept",
	5: "/system/developer",
	6: "/system/broadcast",
	7: "/system/report",
	8: "/crm",
}

func ToCompactToken(bitfield uint64) string {
	return toBase36Upper(bitfield)
}

func MenusFromBitfield(raw string) []string {
	token, ok := ParseSecurityToken(raw)
	if !ok {
		return nil
	}
	var out []string
	for bit, label := range menuBitToToken {
		if hasBitV3(token, bit) {
			out = append(out, label)
		}
	}
	return out
}

func toBase36Upper(value uint64) string {
	if value == 0 {
		return "0"
	}
	const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	var out []byte
	for value > 0 {
		rem := value % 36
		out = append(out, alphabet[rem])
		value /= 36
	}
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return string(out)
}

func ParseSecurityTokenOptional(raw *string) (uint64, bool) {
	if raw == nil {
		return 0, false
	}
	return ParseSecurityToken(*raw)
}

var actionBitToToken = map[int]string{
	actionView:   "view",
	actionCreate: "create",
	actionEdit:   "edit",
	actionDelete: "delete",
	actionExport: "export",
}

func ParseSecurityToken(raw string) (uint64, bool) {
	text := strings.ReplaceAll(strings.TrimSpace(raw), "_", "")
	if text == "" {
		return 0, false
	}
	if rest, ok := strings.CutPrefix(text, "b36:"); ok || strings.HasPrefix(strings.ToUpper(text), "B36:") {
		if !ok {
			rest = text[4:]
		}
		n, ok := parseBase36(rest)
		if ok {
			return normalizeToSingleToken(n), true
		}
	}
	if strings.HasPrefix(strings.ToLower(text), "b64:") {
		rest := text[4:]
		data, err := base64.RawURLEncoding.DecodeString(rest)
		if err != nil {
			data, err = base64.StdEncoding.DecodeString(rest)
		}
		if err == nil && len(data) == 8 {
			var value uint64
			for _, b := range data {
				value = (value << 8) | uint64(b)
			}
			return normalizeToSingleToken(value), true
		}
	}
	if rest, ok := strings.CutPrefix(text, "0x"); ok {
		n, ok := parseHex(rest)
		if ok {
			return normalizeToSingleToken(n), true
		}
	}
	if isAllDigits(text) {
		n, ok := parseBase10(text)
		if ok {
			return normalizeToSingleToken(n), true
		}
	}
	if isAlphaNum(text) {
		n, ok := parseBase36(text)
		if ok {
			return normalizeToSingleToken(n), true
		}
	}
	n, ok := parseBase10(text)
	if ok {
		return normalizeToSingleToken(n), true
	}
	return 0, false
}

func PermissionsFromBitfield(raw string) []string {
	token, ok := ParseSecurityToken(raw)
	if !ok {
		return nil
	}
	var out []string
	for bit, label := range actionBitToToken {
		if hasBitV3(token, bit) {
			out = append(out, label)
		}
	}
	switch ResolveDataScope(token) {
	case "ALL":
		out = append(out, "scope:all")
	case "BRANCH":
		out = append(out, "scope:branch")
	case "DEPARTMENT":
		out = append(out, "scope:department")
	case "OWNER":
		out = append(out, "scope:owner")
	}
	return out
}

func ResolveDataScope(bitfield uint64) string {
	token := normalizeToSingleToken(bitfield)
	scopeMask := extractV3ScopeMask(token)
	if hasMaskBit(scopeMask, v3ScopeAll) {
		return "ALL"
	}
	if hasMaskBit(scopeMask, v3ScopeBranch) {
		return "BRANCH"
	}
	if hasMaskBit(scopeMask, v3ScopeDepartment) {
		return "DEPARTMENT"
	}
	if hasMaskBit(scopeMask, v3ScopeOwner) {
		return "OWNER"
	}
	return "NONE"
}

func BuildBitfield(permissions, menus []string, dev bool) uint64 {
	var menuMask, actionMask, scopeMask uint64
	if dev {
		actionMask = setMaskBit(actionMask, v3ActionView)
		actionMask = setMaskBit(actionMask, v3ActionCreate)
		actionMask = setMaskBit(actionMask, v3ActionEdit)
		actionMask = setMaskBit(actionMask, v3ActionDelete)
		actionMask = setMaskBit(actionMask, v3ActionExport)
		scopeMask = setMaskBit(scopeMask, v3ScopeAll)
	}
	for _, raw := range permissions {
		token := normalizeToken(raw)
		switch token {
		case "dev", "admin":
			actionMask = setMaskBit(actionMask, v3ActionView)
			actionMask = setMaskBit(actionMask, v3ActionCreate)
			actionMask = setMaskBit(actionMask, v3ActionEdit)
			actionMask = setMaskBit(actionMask, v3ActionDelete)
			actionMask = setMaskBit(actionMask, v3ActionExport)
			scopeMask = setMaskBit(scopeMask, v3ScopeAll)
		case "view", "read":
			actionMask = setMaskBit(actionMask, v3ActionView)
		case "create", "add", "insert":
			actionMask = setMaskBit(actionMask, v3ActionCreate)
		case "edit", "update", "write":
			actionMask = setMaskBit(actionMask, v3ActionEdit)
		case "delete", "remove":
			actionMask = setMaskBit(actionMask, v3ActionDelete)
		case "export":
			actionMask = setMaskBit(actionMask, v3ActionExport)
		case "scope:owner", "owner":
			scopeMask = setMaskBit(scopeMask, v3ScopeOwner)
		case "scope:department", "department", "team":
			scopeMask = setMaskBit(scopeMask, v3ScopeDepartment)
		case "scope:branch", "branch":
			scopeMask = setMaskBit(scopeMask, v3ScopeBranch)
		case "scope:all", "all":
			scopeMask = setMaskBit(scopeMask, v3ScopeAll)
		}
	}
	return (menuMask&v3MenuMask)<<v3MenuShift |
		(actionMask&v3ActionMask)<<v3ActionShift |
		(scopeMask&v3ScopeMask)<<v3ScopeShift |
		(v3ReservedSignature & 0xFFFFFFFF)
}

func MergeUniqueCaseInsensitive(base, extra []string) []string {
	seen := make(map[string]struct{})
	var out []string
	for _, value := range append(append([]string{}, base...), extra...) {
		normalized := strings.TrimSpace(value)
		if normalized == "" {
			continue
		}
		key := strings.ToLower(normalized)
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, normalized)
	}
	return out
}

func SubtractCaseInsensitive(base, deny []string) []string {
	denyKeys := make(map[string]struct{})
	for _, value := range deny {
		key := strings.ToLower(strings.TrimSpace(value))
		if key != "" {
			denyKeys[key] = struct{}{}
		}
	}
	var out []string
	for _, value := range base {
		key := strings.ToLower(strings.TrimSpace(value))
		if key == "" {
			continue
		}
		if _, denied := denyKeys[key]; denied {
			continue
		}
		out = append(out, strings.TrimSpace(value))
	}
	return out
}

func HasAdminPrivilege(bitfield uint64) bool {
	token := normalizeToSingleToken(bitfield)
	return hasBitV3(token, actionView) &&
		hasBitV3(token, actionCreate) &&
		hasBitV3(token, actionEdit) &&
		hasBitV3(token, actionDelete) &&
		hasBitV3(token, actionExport) &&
		strings.EqualFold(ResolveDataScope(token), "ALL")
}

func HasActionPermission(permissions []string, action string) bool {
	expected := strings.ToLower(strings.TrimSpace(action))
	if expected == "" {
		return false
	}
	for _, permission := range permissions {
		normalized := strings.ToLower(strings.TrimSpace(permission))
		if normalized == expected || normalized == "admin" {
			return true
		}
	}
	return false
}

func normalizeToken(raw string) string {
	return strings.ToLower(strings.TrimSpace(raw))
}

func normalizeToSingleToken(token uint64) uint64 {
	if isSecurityTokenV3(token) {
		return token
	}
	var menuMask uint64
	for i := 0; i <= 15; i++ {
		if token&(1<<uint(i)) != 0 {
			menuMask = setMaskBit(menuMask, uint64(i))
		}
	}
	var actionMask uint64
	if token&(1<<actionView) != 0 {
		actionMask = setMaskBit(actionMask, v3ActionView)
	}
	if token&(1<<actionCreate) != 0 {
		actionMask = setMaskBit(actionMask, v3ActionCreate)
	}
	if token&(1<<actionEdit) != 0 {
		actionMask = setMaskBit(actionMask, v3ActionEdit)
	}
	if token&(1<<actionDelete) != 0 {
		actionMask = setMaskBit(actionMask, v3ActionDelete)
	}
	if token&(1<<actionExport) != 0 {
		actionMask = setMaskBit(actionMask, v3ActionExport)
	}
	var scopeMask uint64
	if token&(1<<dataScopeOwner) != 0 {
		scopeMask = setMaskBit(scopeMask, v3ScopeOwner)
	}
	if token&(1<<dataScopeDepartment) != 0 {
		scopeMask = setMaskBit(scopeMask, v3ScopeDepartment)
	}
	if token&(1<<dataScopeBranch) != 0 {
		scopeMask = setMaskBit(scopeMask, v3ScopeBranch)
	}
	if token&(1<<dataScopeAll) != 0 {
		scopeMask = setMaskBit(scopeMask, v3ScopeAll)
	}
	return (menuMask&v3MenuMask)<<v3MenuShift |
		(actionMask&v3ActionMask)<<v3ActionShift |
		(scopeMask&v3ScopeMask)<<v3ScopeShift |
		(v3ReservedSignature & 0xFFFFFFFF)
}

func isSecurityTokenV3(token uint64) bool {
	return (token & 0xFFFFFFFF) == (v3ReservedSignature & 0xFFFFFFFF)
}

func hasBitV3(token uint64, bitIndex int) bool {
	if bitIndex < 0 {
		return false
	}
	if bitIndex <= 15 {
		return hasMaskBit(extractV3MenuMask(token), uint64(bitIndex))
	}
	switch bitIndex {
	case actionView:
		return hasMaskBit(extractV3ActionMask(token), v3ActionView)
	case actionCreate:
		return hasMaskBit(extractV3ActionMask(token), v3ActionCreate)
	case actionEdit:
		return hasMaskBit(extractV3ActionMask(token), v3ActionEdit)
	case actionDelete:
		return hasMaskBit(extractV3ActionMask(token), v3ActionDelete)
	case actionExport:
		return hasMaskBit(extractV3ActionMask(token), v3ActionExport)
	case dataScopeOwner:
		return hasMaskBit(extractV3ScopeMask(token), v3ScopeOwner)
	case dataScopeDepartment:
		return hasMaskBit(extractV3ScopeMask(token), v3ScopeDepartment)
	case dataScopeBranch:
		return hasMaskBit(extractV3ScopeMask(token), v3ScopeBranch)
	case dataScopeAll:
		return hasMaskBit(extractV3ScopeMask(token), v3ScopeAll)
	default:
		return false
	}
}

func extractV3MenuMask(token uint64) uint64  { return (token >> v3MenuShift) & v3MenuMask }
func extractV3ActionMask(token uint64) uint64 { return (token >> v3ActionShift) & v3ActionMask }
func extractV3ScopeMask(token uint64) uint64  { return (token >> v3ScopeShift) & v3ScopeMask }

func setMaskBit(mask, bit uint64) uint64 { return mask | (1 << bit) }
func hasMaskBit(mask, bit uint64) bool   { return mask&(1<<bit) != 0 }

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func isAlphaNum(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if (c < '0' || c > '9') && (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') {
			return false
		}
	}
	return true
}

func parseBase10(s string) (uint64, bool) {
	n := new(big.Int)
	if _, ok := n.SetString(s, 10); !ok {
		return 0, false
	}
	return n.Uint64(), true
}

func parseBase36(s string) (uint64, bool) {
	n := new(big.Int)
	if _, ok := n.SetString(strings.ToUpper(s), 36); !ok {
		return 0, false
	}
	return n.Uint64(), true
}

func parseHex(s string) (uint64, bool) {
	n := new(big.Int)
	if _, ok := n.SetString(s, 16); !ok {
		return 0, false
	}
	return n.Uint64(), true
}
